src/json_stream/tokenizer.py

```diff
@@ -299,75 +302,83 @@ def process_char(char, charcode):
                 char = "/"
                 add_char = True
             elif char == "u":
-                next_state = State.UNICODE_1
-                charcode = 0
+                next_state = State.UNICODE
+                unicode_buffer = ""
             else:
                 raise ValueError("Invalid string escape: {}".format(char))
-        elif state == State.UNICODE_1:
-            if char in "0123456789":
-                charcode = (ord(char) - 48) * 4096
-            elif char in "abcdef":
-                charcode = (ord(char) - 87) * 4096
-            elif char in "ABCDEF":
-                charcode = (ord(char) - 55) * 4096
-            else:
-                raise ValueError("Invalid character code: {}".format(char))
-            next_state = State.UNICODE_2
-            char = ""
-        elif state == State.UNICODE_2:
-            if char in "0123456789":
-                charcode += (ord(char) - 48) * 256
-            elif char in "abcdef":
-                charcode += (ord(char) - 87) * 256
-            elif char in "ABCDEF":
-                charcode += (ord(char) - 55) * 256
-            else:
-                raise ValueError("Invalid character code: {}".format(char))
-            next_state = State.UNICODE_3
-            char = ""
-        elif state == State.UNICODE_3:
-            if char in "0123456789":
-                charcode += (ord(char) - 48) * 16
-            elif char in "abcdef":
-                charcode += (ord(char) - 87) * 16
-            elif char in "ABCDEF":
-                charcode += (ord(char) - 55) * 16
+        elif state == State.UNICODE:
+            if char == SpecialChar.EOF:
+                raise ValueError('Unterminated unicode literal at end of file')
+            unicode_buffer += char
+            if len(unicode_buffer) == 4:
+                try:
+                    code_point = int(unicode_buffer, 16)
+                except ValueError:
+                    raise ValueError(f"Invalid unicode literal: \\u{unicode_buffer}")
+                char = chr(code_point)
+                if unicodedata.category(char) == SURROGATE:
+                    next_state = State.UNICODE_SURROGATE_START
+                else:
+                    next_state = State.STRING
+                    add_char = True
+        elif state == State.UNICODE_SURROGATE_START:
+            if char == "\\":
+                next_state = State.UNICODE_SURROGATE_STRING_ESCAPE
+            elif char == SpecialChar.EOF:
+                raise ValueError("Unpaired UTF-16 surrogate at end of file")
```

Comment:
Having all these extra EOF checks is the one thing I'm not sure about. What's the advantage over just outputting the current character (which will read `"EOF"` in case of EOF) from the `else: raise ValueError("Unpaired UTF-16 surrogate")` branches, e.g. raising `ValueError(f"Expected UTF-16 surrogate continuation, got {c!r}")` instead?

Reply:
I just meant that it makes the code quite verbose and wasn't thinking about performance :P But you're right, it should be fine the way it is.